### PR TITLE
Use a Standard File Emission Strategy

### PR DESCRIPTION
**Pull request by davidbiancolin**
Golden Gate generates manyoutput files. Whereas previously this was done in an ad-hoc fashion, here I use FIRRTL's customFileEmission trait to unify how we emit files from various points in the compiler.  Now all output files, with the exception of runtime.conf, will share a common prefix, just as in Rocket Chip's stage. One question is what default prefix should we use? i made it cl_firesim_generated for now. 

Other changes:
- I removed the filtering of the output verilog, and instead use the chisel-emitted macros to do the right thing. 
- I futzed with the names in midas's makefile.

Also i came up with more ideas for simulation-level names:
Metasimulation => Simulating the Simulator 
Fast/Abstract Metasimulation -> MIDAS-level (ML) Sim
Full/Detailed Metasimulation -> FPGA-level sim

TODO:
- [x] Review output filenames and remove in source TODO

#### Related PRs / Issues

This was motivated by my desired to add generated XDC constraints (#800). I suspect there will be more output files in the future. 

firesim/aws-fpga-firesim#42

#### UI / API Impact

Slight output file changes. New names are described in the docs. 

#### Verilog / AGFI Compatibility

It changes the build flow slightly so I regenerated AGFIs. 

### Contributor Checklist
- [x] Did you set dev as the base branch?
- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] (If applicable) Did you regenerate and publicly share default AGFIs?

- [x] (If applicable) Did you mark the PR as "Please Backport"?
- [ ] (On merge) Did you update release notes in the dev-to-master PR ?

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: sim/midas/src/main/cc/Makefile
Comment (davidbiancolin) [inline sim/midas/src/main/cc/Makefile:54]: Unused, i should've rmoved these when i deleted the strober sources.

File: sim/midas/src/main/scala/midas/stage/Annotations.scala
Comment (timsnyder-siv) [inline sim/midas/src/main/scala/midas/stage/Annotations.scala:54]: See, you put Output first in this class name.  I think `OutputFileNameBaseAnnotation` would still be a clearer name.  Related to my first comment on the docs and CLI option name.

File: sim/midas/src/main/scala/midas/stage/Annotations.scala
Comment (timsnyder-siv) [inline sim/midas/src/main/scala/midas/stage/Annotations.scala:59]: Again, why 'most' and not just "all"?  Does it really add something other than potential confusion and having to remember exceptions to the rule to have the runtime.conf exception?

File: sim/midas/src/main/scala/midas/stage/Annotations.scala
Comment (timsnyder-siv) [reply sim/midas/src/main/scala/midas/stage/Annotations.scala:54]: Maybe `OutputBaseFileNameAnnotation` is better because it `OutputFileNameBaseAnnotation` seems to be implying something about a `BaseAnnotation` and that isn't a thing (or maybe it is but it isn't something we're using here [at least not directly] )

Comment (timsnyder-siv) [reply sim/midas/src/main/scala/midas/stage/Annotations.scala:54]: I would make the CLI argument consistent with the name here  so either `--output-base-file-name` or `--output-file-name-base`

File: sim/target-agnostic.mk
Comment (timsnyder-siv) [inline sim/target-agnostic.mk:46]: do you want to do this TODO now or are you intentionally punting for a future PR?

File: sim/Makefile
Comment (timsnyder-siv) [reply sim/Makefile:96]: Nevermind.  I see how `verilog` and `compile` are being moved into `target-agnostic.mk`.

File: sim/target-agnostic.mk
Comment (timsnyder-siv) [inline sim/target-agnostic.mk:186]: s/fields/files/

File: sim/target-agnostic.mk
Comment (timsnyder-siv) [inline sim/target-agnostic.mk:42]: Expand GG to GoldenGate here

**Review by timsnyder-siv — COMMENTED** [review]
I have some questions and other things we should discuss before merging.

File: docs/Golden-Gate/Output-Files.rst
Comment (davidbiancolin) [reply docs/Golden-Gate/Output-Files.rst:5]: Perhaps it's best to just do what RC did?
https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/stage/RocketChipAnnotations.scala#L42

I like prefix personally, but base i used throughout FIRRTL.

File: docs/Golden-Gate/Output-Files.rst
Comment (davidbiancolin) [reply docs/Golden-Gate/Output-Files.rst:6]: Because it was going to be more work... There's another fix i'll fold into this PR and i'll do it. :)

File: docs/Golden-Gate/Output-Files.rst
Comment (davidbiancolin) [reply docs/Golden-Gate/Output-Files.rst:15]: Nearly all drivers take plus args. The runtime_config.ini just encodes a subset of those plus args in an ini format. 

I think a -h would be great. Probably would need to be part of a plus arg rework.

File: docs/Golden-Gate/Output-Files.rst
Comment (davidbiancolin) [reply docs/Golden-Gate/Output-Files.rst:24]: Good idea.

File: sim/midas/src/main/cc/Makefile
Comment (davidbiancolin) [reply sim/midas/src/main/cc/Makefile:10]: There's nothing in this makefile that actually needs to know the _name_ of the top-level class. It's just used to provide a name to things in a variety of places. So really, it's just the name of the target for recipes that also happens to be name of the compiled drivers and simulators. 

I want to get signoff of metasimulator before i commit to it. I mostly dropped it in this PR to try to fish for comments.

File: sim/midas/src/main/scala/midas/stage/Annotations.scala
Comment (davidbiancolin) [reply sim/midas/src/main/scala/midas/stage/Annotations.scala:54]: You should tell me what you think about the RC approach.  Otherwise, i'll switch uniformly to output-base-file-name.

File: sim/midas/src/main/scala/midas/stage/Annotations.scala
Comment (davidbiancolin) [reply sim/midas/src/main/scala/midas/stage/Annotations.scala:59]: runtime.conf. I'll fix it though.

File: sim/target-agnostic.mk
Comment (davidbiancolin) [reply sim/target-agnostic.mk:46]: I'd like to talk about it now, and remove it in this PR. Ill add a box for it.

File: docs/Golden-Gate/Output-Files.rst
Comment (timsnyder-siv) [reply docs/Golden-Gate/Output-Files.rst:5]: Ugh, `--name` and `-n`?  They should probably have just used `--option` and `-o`.  Description much?  I think base or prefix is fine.  Including `output-filename` in the long-name argument is great because it describes what the base or prefix or name is dealing with.

File: docs/Golden-Gate/Output-Files.rst
Comment (timsnyder-siv) [reply docs/Golden-Gate/Output-Files.rst:15]: But I could add any of the plusargs to `runtime.conf` (as it works right now because it just gets catted into the launch cmd), right?

Comment (davidbiancolin) [reply docs/Golden-Gate/Output-Files.rst:15]: correctomundo

Comment (davidbiancolin) [reply docs/Golden-Gate/Output-Files.rst:15]: @sagark was going to add a passthrough in the config runtime.ini, that would essentially be a list of additional plus args

File: sim/midas/src/main/cc/Makefile
Comment (timsnyder-siv) [reply sim/midas/src/main/cc/Makefile:10]: Oh, yeah, would help if I looked at the context before writing my comment.  This is just in the driver Makefile and the toplevel makefile still uses DESIGN and that does need to be the toplevel classname because it's what is used in the generator invocation?  Gotcha.  This makes sense.

File: sim/midas/src/main/scala/midas/stage/Annotations.scala
Comment (timsnyder-siv) [reply sim/midas/src/main/scala/midas/stage/Annotations.scala:54]: Switch to `output-base-file-name`.  If GH had the vomit emoji, I'd use it to express how I feel about the RC option name. 🤮   Oh, it does have the 🤮 emoji!  yay.  That's how I feel about it.

File: sim/firesim-lib/src/main/scala/passes/EC2F1Artefacts.scala
Comment (davidbiancolin) [reply sim/firesim-lib/src/main/scala/passes/EC2F1Artefacts.scala:28]: > I'm fairly certain that Vivado defines SYNTHESIS by default. I think we talked about this last year somewhere after I had done some searching to confirm that it does. I'll look that up again.

Oh right, true. I'll just remove it here then. 

But no, it is not included.  Maybe i could give it a different, better name perhaps, synth_defines.vh ?

File: sim/firesim-lib/src/main/scala/passes/ILATopWiring.scala
Comment (davidbiancolin) [reply sim/firesim-lib/src/main/scala/passes/ILATopWiring.scala:42]: Yeah, the references are tricky. Maybe just a URL to latest is best? Or even a fixed version?

File: sim/firesim-lib/src/main/scala/passes/ILATopWiring.scala
Comment (davidbiancolin) [reply sim/firesim-lib/src/main/scala/passes/ILATopWiring.scala:159]: Nope, that's a mistake. I should use it below in place of wiredState.annotations.

File: docs/Golden-Gate/Output-Files.rst
Comment (davidbiancolin) [inline docs/Golden-Gate/Output-Files.rst:14]: These names need to be updated.

File: sim/midas/src/main/scala/midas/passes/SimulationMapping.scala
Comment (davidbiancolin) [inline sim/midas/src/main/scala/midas/passes/SimulationMapping.scala:115]: I reordered this so that module definitions precede instantiations.

File: sim/midas/src/main/scala/midas/passes/ElaborateChiselSubCircuit.scala
Comment (davidbiancolin) [inline sim/midas/src/main/scala/midas/passes/ElaborateChiselSubCircuit.scala:13]: I wrote this to replace the flow in SimulationMapping so that FileAnnotations generated from Chisel in the simulator wrapper would get properly serialized out.

File: sim/midas/src/main/scala/midas/stage/GoldenGateCompilerPhase.scala
Comment (davidbiancolin) [inline sim/midas/src/main/scala/midas/stage/GoldenGateCompilerPhase.scala:43]: This emitter is essentially a wrapper around the VerilogEmitter. However, since we append black boxes that do have SV constructs it was simplest to simply use this emitter and get the desired .sv suffix immediately. (You can't use the Verilog emitter and ask for a "<name>.sv" output file...)

File: sim/src/main/makefrag/fasedtests/Makefrag
Comment (davidbiancolin) [inline sim/src/main/makefrag/fasedtests/Makefrag:29]: These are no longer set by the Makefrags, and instead are derived from BASE_FILE_NAME.

File: sim/target-agnostic.mk
Comment (davidbiancolin) [inline sim/target-agnostic.mk:12]: I just reordered these a little bit to group them better by function

File: sim/target-agnostic.mk
Comment (davidbiancolin) [inline sim/target-agnostic.mk:56]: I renamed HEADER -> header, VERILOG -> simulator_verilog, to try to maintain that uppercase variables are overridable, whilst lowercase ones are not. I prepended simulator_ to avoid a confusion collision with the phony target.

File: sim/target-agnostic.mk
Comment (davidbiancolin) [inline sim/target-agnostic.mk:172]: I don't enumerate individual files anymore because they can cleanly be copied en masse (since they have consistent names).

File: sim/target-agnostic.mk
Comment (davidbiancolin) [inline sim/target-agnostic.mk:198]: This is now set with a verilog `define. And `SYNTHESIS hands the other case.

Comment (davidbiancolin) [reply sim/target-agnostic.mk:198]: This has the nice property of no longer needing to explicitly modify the output verilog for Vivado.

**Review by davidbiancolin — COMMENTED** [review]
Self-review trying to explain my thinking.

File: sim/target-agnostic.mk
Comment (timsnyder-siv) [reply sim/target-agnostic.mk]: Did you do this?  I see the box you added in the PR description and it's still unchecked.

**Review by timsnyder-siv — COMMENTED** [review]
I think there's still a few things to discuss a bit before I approve and this gets merged.

File: sim/midas/src/main/scala/midas/passes/ElaborateChiselSubCircuit.scala
Comment (davidbiancolin) [reply sim/midas/src/main/scala/midas/passes/ElaborateChiselSubCircuit.scala:15]: Im simulation mapping there was an ad-hoc chunk of code that called the chisel stage. This replaces that. I'll make sure that's clear in this comment.

File: sim/target-agnostic.mk
Comment (davidbiancolin) [reply sim/target-agnostic.mk:148]: Correct. Anything "chain" related.

File: sim/midas/src/main/scala/midas/stage/GoldenGateFileEmission.scala
Comment (davidbiancolin) [reply sim/midas/src/main/scala/midas/stage/GoldenGateFileEmission.scala:36]: It looks like this: https://github.com/firesim/firesim/pull/802/files#diff-43e77e02a0cf4dfa093ddf410a9ac665b16f8be6c7e1eec75cdc790155728517R269

Frankly i detest the pattern of explicitly naming Chisel annotation classes that essentially just wrap a FIRRTL annotation. Instead, let's get the name of the method right, and possibly put it in an object with a better name (then it could just be an apply method).

File: sim/firesim-lib/src/main/scala/passes/ILATopWiring.scala
Comment (timsnyder-siv) [reply sim/firesim-lib/src/main/scala/passes/ILATopWiring.scala:162]: looks good now.

File: sim/midas/src/main/scala/midas/stage/GoldenGateFileEmission.scala
Comment (timsnyder-siv) [reply sim/midas/src/main/scala/midas/stage/GoldenGateFileEmission.scala:36]: > Frankly i detest the pattern of explicitly naming Chisel annotation ...

fair.  And, this is probably the strongest reaction I've ever seen out of you.  Take it easy... ;)

**Review by timsnyder-siv — APPROVED** [review]

**Pull request merged**
